extensions/amp-access/0.1/access-expr.js

```diff
@@ -44,3 +45,51 @@ export function evaluateAccessExpr(expr, data) {
     parser.yy = null;
   }
 }
+
+/**
+ * AmpAccessEvaluator evaluates amp-access expressions.
+ * It uses a cache to speed up repeated evals for the same expression.
+ */
+export class AmpAccessEvaluator {
+  /** */
+  constructor() {
+    /** @type {Object<string, boolean>} */
+    this.cache = null;
+
+    /** @type {JsonObject} */
+    this.lastData = null;
+  }
+
+  /**
+   * Evaluate access expressions, but turn to a cache first.
+   * Cache is invalidated when given new data.
+   *
+   * @param {string} expr
+   * @param {!JsonObject} data
+   * @return {boolean}
+   */
+  evaluate(expr, data) {
+    if (this.lastData !== data) {
```

Comment:
This will work for the  first time through when data was null but if the data object changes attributes internally I don't think this will catch it which could be an issue with amp-subscriptions - I need to check how we merge objects to be sure but I'm 90% sure we just add the new attributes which this won't catch.

Reply:
Interesting. I realized I wasn't sure if this was also a bug for amp-access so I just double checked and it looks like `amp-access.js` creates a new object each time it kicks off a chain of authz calls. So that should be fine.

For `amp-subscriptions`, do you have any example of a site that may call the amp-access parser many times? If yes I can run some measurements

Reply:
The edge case I was worried about is when you have amp-subscriptions and amp-subscriptions-google installed and a potential race.  Looking through the code I think we're ok at the moment, we use the entitlement.json() which is rebuilt every time it's called.  I think you're safe to add it to amp-subscriptions, maybe with a few strong comments explaining what's going.

Reply:
I'd rather hold off on adding this perf optimization to `amp-subscriptions` at least until a future PR where I can spend more time ensuring it is safe and that it actually speeds up websites.